### ISSUE TITLE
[NEXT-876] MODULE_NOT_FOUND in Vercel production/preview environments, but not locally

### ISSUE THREAD

**Issue by zeke** [vercel/next.js#47380]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
$ next info              

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:04:44 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.13.0
      npm: 8.19.3
      Yarn: 1.22.15
      pnpm: 7.27.0
    Relevant packages:
      next: 13.2.5-canary.10
      eslint-config-next: 12.2.5
      react: 18.2.0
      react-dom: 18.2.0

warn  - Latest canary version not detected, detected: "13.2.5-canary.10", newest: "13.2.5-canary.11".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```

:point_up: Note there's a warning here about me not using the latest canary, but `npm i next@canary` installs 10 for me, not 11.

### Which area(s) of Next.js are affected? (leave empty if unsure)

Package manager (npm, pnpm, Yarn)?

### Link to the code that reproduces this issue

https://github.com/replicate/scribble-diffusion/pull/46

### To Reproduce

Run it locally with success:

```sh
git clone https://github.com/replicate/scribble-diffusion
gh pr checkout 46
export REPLICATE_API_TOKEN="..." # from https://replicate.com/account
npm install
npm run dev
```

Go to http://localhost:3000, hit the Go button, and watch it work. 👍🏼 

---

Push to Vercel and watch it fail:

```
vercel
vercel logs -f YOUR_DEPLOYMENT_URL
```

Go to your deployment URL and hit the "Go" button. You'll see an error in the browser console and the tailed server logs. 😿 

### Describe the Bug

I am using the `replicate` npm package in my Next.js application. It works locally, but produces this error in production (and in review apps) at runtime:

```
2023-03-13T19:06:25.472Z  2023-03-13T19:06:26.632Z	4c7e6aa0-c7c4-4f91-b02c-e0f56b9f7e38	ERROR	Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/node_modules/replicate/lib/errors.js' imported from /var/task/node_modules/replicate/lib/ReplicateClient.js
2023-03-13T19:06:25.472Z      at new NodeError (node:internal/errors:400:5)
2023-03-13T19:06:25.472Z      at finalizeResolution (node:internal/modules/esm/resolve:331:11)
2023-03-13T19:06:25.472Z      at moduleResolve (node:internal/modules/esm/resolve:994:10)
2023-03-13T19:06:25.472Z      at moduleResolveWithNodePath (node:internal/modules/esm/resolve:938:12)
2023-03-13T19:06:25.472Z      at defaultResolve (node:internal/modules/esm/resolve:1202:79)
2023-03-13T19:06:25.472Z      at nextResolve (node:internal/modules/esm/loader:163:28)
2023-03-13T19:06:25.472Z      at ESMLoader.resolve (node:internal/modules/esm/loader:842:30)
2023-03-13T19:06:25.472Z      at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
2023-03-13T19:06:25.472Z      at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
2023-03-13T19:06:25.472Z      at link (node:internal/modules/esm/module_job:76:36) {
2023-03-13T19:06:25.472Z    code: 'ERR_MODULE_NOT_FOUND'
2023-03-13T19:06:25.472Z  }
```

Related bug: https://github.com/replicate/replicate-js/issues/24

Notes:

- It works locally with `npm run dev` as well as `npm run build && npm start`.
- The offending code is in an [API (server-side) route](https://github.com/replicate/scribble-diffusion/blob/0bfbaa085150b7159fbf987b9e2969080814520c/pages/api/predictions/index.js#L1), so I would not expect webpack/browser bundling to be to blame here?
- The build doesn't fail. It's not until I invoke the server-side route that the error occurs.

Things I have tried:

- Pinning the Node.js version in `engines.node` so I know I'm using the same thing locally and on Vercel: https://github.com/replicate/scribble-diffusion/pull/46/commits/221f4722d7087fd585ca929d4bc9a61d72401aff
- Bumping `next` to the latest version (and the canary): https://github.com/replicate/scribble-diffusion/pull/46/commits/899269d75e0d8cb940faf9b84dc128a5a3dc8a7c
- Disabling `esmExternals`: https://github.com/replicate/scribble-diffusion/pull/46/commits/e0d7dcab11a2347a9e52881f5acbd5b73aa59053
- Avoiding using the new relatively new import/assert syntax: https://github.com/replicate/scribble-diffusion/pull/46/commits/4eb5af97225a9af645965ed4d82a3fb00fb543d2

### Expected Behavior

There should not be a `MODULE_NOT_FOUND` error in production.

I should be able to reproduce the error locally somehow.

### How are you deploying your application? (if relevant)

Vercel's GitHub integration

<sub>[NEXT-876](https://linear.app/vercel/issue/NEXT-876/module-not-found-in-vercel-productionpreview-environments-but-not)</sub>

**Comment by balazsorban44** [comment]
It seems to be related to the import assertion on the line before:

https://github.com/replicate/replicate-js/blob/6fa9006f7dae32d3b04d1a4947782aacfceef844/lib/ReplicateClient.js#L1

Using `patch-package` and removing it makes it work. :thinking:

I'm more curious about how it works locally. It is a Stage 2 proposal as far as I can tell:
https://github.com/tc39/proposal-import-assertions
https://tc39.es/proposal-import-assertions

I will check with the team

**Comment by zeke** [comment]
Hi @balazsorban44 thanks for looking into it. That import/assert line did seem suspicious so I tried removing it: 

https://github.com/replicate/scribble-diffusion/commit/4eb5af97225a9af645965ed4d82a3fb00fb543d2
https://github.com/replicate/replicate-js/commit/1496e3000c334d09c49c6229749c2d3b55b302d8

But that didn't seem to fix the issue.

**Comment by jridgewell** [comment]
> I'm more curious about how it works locally. It is a Stage 2 proposal as far as I can tell:

Newly stage 2 in Jan, before that it was Stage 3 since 2020-09. It's actually going back to Stage 3 after today's meeting.

**Comment by zeke** [comment]
Got it sorted. Thanks @balazsorban44 and @jridgewell for your help on this. 🙏🏼
